File: testng-core/src/test/java/org/testng/internal/ClassHelperTest.java
package org.testng.internal;

import static org.assertj.core.api.Assertions.assertThat;

import java.lang.reflect.Method;
import java.net.URL;
import java.net.URLClassLoader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Set;
import java.util.UUID;
import org.assertj.core.api.Assertions;
import org.testng.Assert;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;
import org.testng.internal.issue1339.BabyPanda;
import org.testng.internal.issue1339.LittlePanda;
import org.testng.internal.issue1456.TestClassSample;
import org.testng.internal.misamples.AbstractMoves;
import org.testng.internal.misamples.Batman;
import org.testng.internal.misamples.JohnTravoltaMoves;
import org.testng.internal.misamples.MickJagger;
import org.testng.xml.XmlClass;
import org.testng.xml.XmlSuite;
import org.testng.xml.XmlTest;

public class ClassHelperTest {

  @Test
  public void testGetAvailableMethods() {
    runTest(getExpected(), LittlePanda.class);
  }

  @Test
  public void testGetAvailableMethodsWhenOverrdingIsInvolved() {
    List<String> expected = getExpected("equals", "hashCode", "toString");
    runTest(expected, BabyPanda.class);
  }

  @Test
  public void testFindClassInSameTest() {
    runTest(TestClassSample.class, 1, TestClassSample.class);
  }

  @Test
  public void testFindClassesInSameTest() {
    runTest(TestClassSample.class, 2, TestClassSample.class, BabyPanda.class);
  }

  @Test
  public void testNoClassDefFoundError() {
    URLClassLoader urlClassLoader =
        new URLClassLoader(new URL[] {}) {
          @Override
          public Class<?> loadClass(String name) throws ClassNotFoundException {
            throw new NoClassDefFoundError();
          }
        };
    ClassHelper.addClassLoader(urlClassLoader);
    String fakeClassName = UUID.randomUUID().toString();
    Assert.assertNull(
        ClassHelper.forName(fakeClassName),
        "The result should be null; no exception should be thrown.");
  }

  @Test(dataProvider = "data")
  public void testWithDefaultMethodsBeingOverridden(
      Class<?> cls, int expectedCount, String... expected) {
    Set<Method> methods = ClassHelper.getAvailableMethodsExcludingDefaults(cls);
    Assertions.assertThat(methods).hasSize(expectedCount);
    for (Method m : methods) {
      String actual = m.getDeclaringClass().getName() + "." + m.getName();
      Assertions.assertThat(expected).contains(actual);
    }
  }

  @DataProvider(name = "data")
  public Object[][] getTestData() {
    return new Object[][] {
      {MickJagger.class, 1, MickJagger.class.getName() + ".dance"},
      {
        JohnTravoltaMoves.class,
        2,
        new String[] {
          JohnTravoltaMoves.class.getName() + ".walk", AbstractMoves.class.getName() + ".dance"
        }
      },
      {
        Batman.class,
        3,
        new String[] {
          Batman.class.getName() + ".fly",
          Batman.class.getName() + ".liftWeights",
          Batman.class.getName() + ".yellSlogan"
        }
      }
    };
  }

  private static void runTest(Class<?> classToBeFound, int expectedCount, Class<?>... classes) {
    XmlSuite xmlSuite = new XmlSuite();
    xmlSuite.setName("xml_suite");
    newXmlTest("test1", xmlSuite, classes);
    newXmlTest("test2", xmlSuite, classes);
    newXmlTest("test3", xmlSuite, classes);
    XmlClass[] xmlClasses = ClassHelper.findClassesInSameTest(classToBeFound, xmlSuite);
    assertThat(xmlClasses.length).isEqualTo(expectedCount);
  }

  private static void newXmlTest(String testname, XmlSuite xmlSuite, Class<?>... clazz) {
    XmlTest xmlTest = new XmlTest(xmlSuite);
    xmlTest.setName(testname);
    xmlTest.setXmlClasses(newXmlClass(clazz));
  }

  private static List<XmlClass> newXmlClass(Class<?>... classes) {
    List<XmlClass> xmlClasses = new ArrayList<>();
    for (Class<?> clazz : classes) {
      xmlClasses.add(new XmlClass(clazz));
    }
    return xmlClasses;
  }

  private static void runTest(List<String> expected, Class<?> whichClass) {
    Set<Method> methods = ClassHelper.getAvailableMethods(whichClass);
    // Intentionally not using assertEquals because when this test is executed via gradle an
    // additional method
    // called "jacocoInit()" is getting added, which does not get added when this test is executed
    // individually
    int size = expected.size();
    Assert.assertTrue(
        methods.size() >= size, "Number of methods found should have been atleast " + size);
    for (Method method : methods) {
      if ("$jacocoInit".equalsIgnoreCase(method.getName())) {
        continue;
      }
      Assert.assertTrue(expected.contains(method.getName()));
    }
  }

  private static List<String> getExpected(String... additionalMethods) {
    String[] defaultMethods = new String[] {"announcer", "announcer", "inheritable", "inheritable"};
    if (additionalMethods == null) {
      return Arrays.asList(defaultMethods);
    }
    List<String> expected = new ArrayList<>(Arrays.asList(defaultMethods));
    expected.addAll(Arrays.asList(additionalMethods));
    return expected;
  }
}

File: testng-core/src/test/kotlin/org/testng/internal/ClassHelperTest.kt
package org.testng.internal

import org.assertj.core.api.Assertions.assertThat
import org.testng.Reporter

import java.net.URL
import java.net.URLClassLoader
import java.util.UUID
import org.testng.annotations.DataProvider
import org.testng.annotations.Test
import org.testng.internal.samples.classhelper.issue1339.BabyPandaSample
import org.testng.internal.samples.classhelper.issue1339.LittlePandaSample
import org.testng.internal.samples.classhelper.issue1456.TestClassSample
import org.testng.internal.samples.classhelper.misamples.AbstractMovesSample
import org.testng.internal.samples.classhelper.misamples.BatmanSample
import org.testng.internal.samples.classhelper.misamples.JohnTravoltaMovesSample
import org.testng.internal.samples.classhelper.misamples.MickJaggerSample
import org.testng.xml.XmlClass
import org.testng.xml.XmlSuite
import org.testng.xml.XmlTest
import kotlin.reflect.KClass

typealias TestData = Array<Array<Any>>

class ClassHelperTest {

    private val defaultMethods = listOf("announcer", "announcer", "inheritable", "inheritable")

    @Test(dataProvider = "testDataProvider")
    fun runFindClassesInSameTestScenarios(
        scenarioName: String,//This parameter will help us differentiate scenarios in reports
        classToBeFound: KClass<*>,
        expectedCount: Int, vararg classes: KClass<*>,
    ) {
        Reporter.log("Running Scenario $scenarioName")
        val xmlSuite = XmlSuite().apply {
            name = "xml_suite"
        }
        listOf("test1", "test2", "test3").forEach {
            xmlSuite.addXmlTest(it, *classes)
        }
        val xmlClasses = ClassHelper.findClassesInSameTest(classToBeFound.java, xmlSuite)
        assertThat(xmlClasses).hasSize(expectedCount)
    }

    @DataProvider(name = "testDataProvider")
    fun testDataProvider(): TestData {
        return arrayOf(
            arrayOf("FindClassInSameTest", TestClassSample::class, 1, TestClassSample::class),
            arrayOf(
                "FindClassesInSameTest",
                TestClassSample::class, 2, TestClassSample::class, BabyPandaSample::class
            ),
        )
    }

    @Test(dataProvider = "scenariosData")
    fun runAvailableMethodsScenarios(
        scenarioName: String,//This parameter will help us differentiate scenarios in reports
        expected: List<String>,
        whichClass: KClass<*>
    ) {
        Reporter.log("Running Scenario $scenarioName")
        val actual = ClassHelper.getAvailableMethods(whichClass.java)
            .filter {
                //jacocoInit() is something that Gradle adds. So filter it out.
                !"\$jacocoInit".contentEquals(it.name)
            }
            .map {
                it.name
            }.toList()
        assertThat(actual).containsAll(expected)
    }

    @DataProvider(name = "scenariosData")
    fun scenariosData(): TestData {
        return arrayOf(
            arrayOf("AllAvailableMethods", defaultMethods, LittlePandaSample::class),
            arrayOf(
                "AvailableMethodsWhenOverridingIsInvolved",
                defaultMethods + listOf("equals", "hashCode", "toString"),
                BabyPandaSample::class
            ),
        )
    }

    @Test
    fun testNoClassDefFoundError() {
        val urlClassLoader = object : URLClassLoader(emptyArray<URL>()) {
            override fun loadClass(className: String?): Class<*> {
                throw NoClassDefFoundError()
            }
        }
        ClassHelper.addClassLoader(urlClassLoader)
        val fakeClassName = UUID.randomUUID().toString()
        assertThat(ClassHelper.forName(fakeClassName))
            .withFailMessage("The result should be null; no exception should be thrown.")
            .isNull()
    }

    @Test(dataProvider = "data")
    fun testWithDefaultMethodsBeingOverridden(
        cls: KClass<*>, expectedCount: Int, vararg expected: String
    ) {
        val methods = ClassHelper.getAvailableMethodsExcludingDefaults(cls.java)
            .map {
                it.declaringClass.name + "." + it.name
            }
        assertThat(methods).hasSize(expectedCount)
        assertThat(methods).contains(*expected)
    }

    @DataProvider(name = "data")
    fun getTestData(): TestData {
        return arrayOf(
            arrayOf(MickJaggerSample::class, 1, MickJaggerSample::class.java.name + ".dance"),
            arrayOf(
                JohnTravoltaMovesSample::class,
                2,
                JohnTravoltaMovesSample::class.java.name + ".walk",
                AbstractMovesSample::class.java.name + ".dance"
            ),
            arrayOf(
                BatmanSample::class,
                3,
                BatmanSample::class.java.name + ".fly",
                BatmanSample::class.java.name + ".liftWeights",
                BatmanSample::class.java.name + ".yellSlogan"
            )
        )
    }

    private fun XmlSuite.addXmlTest(testName: String, vararg clazz: KClass<*>): XmlSuite {
        XmlTest(this).apply {
            name = testName
            xmlClasses = newXmlClasses(*clazz)
        }
        return this
    }

    private fun newXmlClasses(vararg classes: KClass<*>): List<XmlClass> {
        return classes.map {
            XmlClass(it.java)
        }.toList()
    }
}
